circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,471 @@
+	for LookupShortWordCheckChip<F, K, S>
+{
+	type Output = AssignedCell<F, F>;
+
+	fn configure(
+		mock_common: &MockCommonConfig, meta: &mut halo2::plonk::ConstraintSystem<F>,
+	) -> Selector {
+		assert!(0 < S && S < K, "Word bits should be less than target bits.");
+
+		let bitshift_selector = meta.complex_selector();
+
+		let word_column = mock_common.common.advice[0];
+
+		meta.lookup("Check K bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let shifted_word = meta.query_advice(word_column, Rotation::cur());
+
+			vec![(bitshift_selector * shifted_word, mock_common.table)]
+		});
+
+		// For short lookups, check that the word has been shifted by the correct number
+		// of bits. https://p.z.cash/halo2-0.1:decompose-short-lookup
+		meta.create_gate("Short word S bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let word = meta.query_advice(word_column, Rotation::prev());
+			let shifted_word = meta.query_advice(word_column, Rotation::cur());
+			let inv_two_pow_s = meta.query_advice(word_column, Rotation::next());
+
+			let two_pow_k = F::from(1 << K);
+
+			// shifted_word = word * 2^{K-s}
+			//              = word * 2^K * inv_two_pow_s
+			Constraints::with_selector(
+				bitshift_selector,
+				Some(word * two_pow_k * inv_two_pow_s - shifted_word),
+			)
+		});
+
+		bitshift_selector
+	}
+
+	fn synthesize(
+		self, mock_common: &MockCommonConfig, selector: &Selector, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, halo2::plonk::Error> {
+		layouter.assign_region(
+			|| "short word check chip",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+
+				// Assign original value
+				let assigned_x = ctx.copy_assign(mock_common.common.advice[0], self.x.clone())?;
+
+				ctx.next();
+
+				// Assign shifted value
+				let shifted_word = self.x.value().cloned() * Value::known(F::from(1 << (K - S)));
+				ctx.assign_advice(mock_common.common.advice[0], shifted_word)?;
+				ctx.enable(selector.clone())?;
+
+				ctx.next();
+
+				// Assign 2^{-S} from a fixed column.
+				let inv_two_pow_s = F::from(1 << S).invert().unwrap();
+				ctx.assign_from_constant(mock_common.common.advice[0], inv_two_pow_s)?;
+
+				Ok(assigned_x)
+			},
+		)
+	}
+}
+
+/// Range check chip using lookup table.
+///
+/// Check if the target value is `K * N` bits long.
+#[derive(Debug, Clone)]
+pub struct LookupRangeCheckChip<F: FieldExt, const K: usize, const N: usize> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> LookupRangeCheckChip<F, K, N> {
+	/// Construct a new instance
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> MockChip<F> for LookupRangeCheckChip<F, K, N> {
+	type Output = AssignedCell<F, F>;
+
+	fn configure(
+		mock_common: &MockCommonConfig, meta: &mut halo2::plonk::ConstraintSystem<F>,
```

Comment:
We should import `ConstraintSystem` at the top of the file, instead of specifying the whole path:  `halo2::plonk::ConstraintSystem`
Do the same thing in every similar situation

Reply:
Done https://github.com/eigen-trust/protocol/pull/169/commits/f5cb0f482c9f45b0327720226205386fe8a03613